api/v1beta1/types.go

```diff
@@ -107,6 +107,161 @@ type Network struct {
+	// required when creating a firewall rule. This value can either be one of the
+	// following well known protocol strings (tcp, udp, icmp, esp, ah, ipip, sctp)
+	// or the IP protocol number.
+	IPProtocol string `json:"IPProtocol,omitempty"`
+	// Ports: An optional list of ports to which this rule applies. This field is
+	// only applicable for the UDP or TCP protocol. Each entry must be either an
+	// integer or a range. If not specified, this rule applies to connections
+	// through any port. Example inputs include: ["22"], ["80","443"], and
+	// ["12345-12349"].
+	Ports []string `json:"ports,omitempty"`
+}
+
+// FirewallRule describes a GCP firewall rule.
+type FirewallRule struct {
+	// Allowed: The list of ALLOW rules specified by this firewall. Each rule
+	// specifies a protocol and port-range tuple that describes a permitted
+	// connection.
+	Allowed []*FirewallDescriptor `json:"allowed,omitempty"`
+	// Denied: The list of DENY rules specified by this firewall. Each rule
+	// specifies a protocol and port-range tuple that describes a denied
+	// connection.
+	Denied []*FirewallDescriptor `json:"denied,omitempty"`
+	// Description: An optional description of this resource. Provide this field
+	// when you create the resource.
+	Description *string `json:"description,omitempty"`
+	// DestinationRanges: If destination ranges are specified, the firewall rule
+	// applies only to traffic that has destination IP address in these ranges.
+	// These ranges must be expressed in CIDR format. Both IPv4 and IPv6 are
+	// supported.
+	DestinationRanges []string `json:"destinationRanges,omitempty"`
+	// Direction: Direction of traffic to which this firewall applies, either
+	// `INGRESS` or `EGRESS`. The default is `INGRESS`. For `EGRESS` traffic, you
+	// cannot specify the sourceTags fields.
+	//
+	// Possible values:
+	//   "EGRESS" - Indicates that firewall should apply to outgoing traffic.
+	//   "INGRESS" - Indicates that firewall should apply to incoming traffic.
+	// +kubebuilder:validation:Enum=INGRESS;EGRESS
+	// +kubebuilder:default=INGRESS
+	// +optional
+	Direction *FirewallRuleDirection `json:"direction,omitempty"`
+	// Disabled: Denotes whether the firewall rule is disabled. When set to true,
+	// the firewall rule is not enforced and the network behaves as if it did not
+	// exist. If this is unspecified, the firewall rule will be enabled.
+	Disabled *bool `json:"disabled,omitempty"`
+	// Name: Name of the resource; provided by the client when the resource is
+	// created. The name must be 1-63 characters long, and comply with RFC1035.
+	// Specifically, the name must be 1-63 characters long and match the regular
+	// expression `[a-z]([-a-z0-9]*[a-z0-9])?`. The first character must be a
+	// lowercase letter, and all following characters (except for the last
+	// character) must be a dash, lowercase letter, or digit. The last character
+	// must be a lowercase letter or digit.
+	Name *string `json:"name,omitempty"`
```

Comment:
Thinking about collisions here - should we use this as a suffix with the cluster name?